infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -292,13 +298,24 @@ private void rangeWrite(CorfuPayloadMsg<RangeWriteMsg> msg,
     }
 
     /**
-     * Resets the log unit server.
+     * Resets the log unit server via the BatchWriter.
      * Warning: Clears all data.
      */
     @ServerHandler(type = CorfuMsgType.RESET_LOGUNIT)
-    private void resetLogUnit(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        streamLog.reset();
-        dataCache.invalidateAll();
+    private synchronized void resetLogUnit(CorfuPayloadMsg<Long> msg,
+                                           ChannelHandlerContext ctx, IServerRouter r) {
+        // Check if the reset request is with an epoch greater than the last reset epoch seen.
+        // and should be equal to the current router epoch to prevent stale reset requests from
+        // wiping out the data.
+        if (msg.getPayload() > serverContext.getLogUnitResetEpoch()
+                && msg.getPayload() == serverContext.getServerEpoch()) {
+            batchWriter.reset();
+            dataCache.invalidateAll();
+            serverContext.setLogUnitResetEpoch(msg.getPayload());
+            log.info("LogUnit Server Reset.");
+        } else {
+            log.info("LogUnit Server Reset request received but reset already done.");
+        }
```

Comment:
I think this has a race condition. If the batch is at a state where the queue looks like [Write, Reset, Write], then you'll end up with data after the reset. It needs to be an atomic operation.

Reply:
We do a seal to prevent all writes in the current epoch and then perform a reset.
So any writes in the queue after the reset will be from the next epoch.

Reply:
That might not be true. The state of the queue doesn't reflect all pending operations. For example, a reading thread might block a write thread on the cache itself. Before the write thread is unblocked, a reset request comes in, the queue can still have a write after the reset.